posts/change-data-capture-cloudsql-postgres-with-apache-airflow-1.md

```diff
@@ -0,0 +1,1002 @@
+
+The `execution_date` is a `pendulum.Pendulum` object available via a jinja templated macro as `{{ execution_date }}`. It's defined as a standard time relative to that which the DAG was scheduled to run at, independent of when it started or was re-ran.
+
+All Airflow macros, including `execution_date`, can be used to template our SQL command. With that, we can begin by letting the DAG `execution_date` be the `high_watermark` and subtract an hour for the `low_watermark`.
+
+Essentially, a basic version of the templated SQL query might look something like this:
+
+```sql
+{%- set low_watermark =  execution_date.subtract(hours=1) -%}
+SELECT * FROM some_schema.table_1
+WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
+AND event_timestamp AT TIME ZONE 'UTC' < '{{ execution_date }}'
+```
+
+If you have done this before, you might be thinking we will need to adjust for how the `execution_date` is calculated relative to the schedule interval. We will do that in `Step 12`. First, we need to template our SQL. To do so:
+
+1. Make some SQL files formatted as `table_name.sql` , as well as a `get_schema.sql`file, in the following path:
+
+    ```bash
+    ├── include
+    │   └── templates
+    │       ├── cloud_sql_cdc
+    │       │   └── export_csv
+    │       │       ├── table_1.sql
+    │       │       ├── table_2.sql
+    │       │       ├── table_3.sql
+    │       │       ├── table_4.sql
+    │       │       ├── get_schema.sql
+    ```
+
+ 2. Add the following to each `table_{num}.sql` file you created, changing the table name to your own, accordingly:
+    ```sql
+    {%- set low_watermark =  execution_date.subtract(hours=1) -%}
+    SELECT * FROM some_schema.table_1
+    WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
+    AND event_timestamp AT TIME ZONE 'UTC' < '{{ execution_date }}'
+    ```
+
+3. Combine all tables into a single schema CSV by adding the following to the  `get_schema.sql` (and changing `some_schema`):
+    ```sql
+    {% set tables = params.tables %}
+    SELECT 
+      table_name,
+      column_name,
+      data_type,
+      ordinal_position
+    FROM information_schema.columns 
+    WHERE table_schema = 'some_schema' 
+    AND table_name in (
+      {%- for name in tables -%}
+        {%- if loop.first -%}
+        '{{ name }}'
+        {%- else -%}
+        ,'{{ name }}'
+        {%- endif -%}
+      {%- endfor -%}
+    )
+    ```
+
+    The `ordinal_position` field is how you will be able to assign column names to header fields in the CSV file. The `params.tables` assignment will come in the DAG file. Note, it is also possible to join tables here (in the need for immediate denormalization) by a `UNION SELECT` on this to the join query with an assigned table name. Now we're able to include functionality in our Airflow tasks.
```

Comment:
Think this might end up throwing some markdown errors because only the first line is tabbed out, but check it out in your preview.